### PR TITLE
Pull and push Rigid constraints

### LINKED ISSUE #296: GSA_Toolkit: Pull and Push Constraints
#### Description:

Add functionality to push and pull constraints.

### PR DESCRIPTION
**Pull request by TosteSkDa**
Added feature to read, pull and push rigid constraints from/to GSA. RigidConstraints are treated by BHoM as a RigidLink but with a added Fragment that tells GSA that it is a RigCon instead. Previously added unique methods for RigCons have been removed and this is now an adjustment to existing Rigid Link methods. 

 ### NOTE: Depends on 

 ### Issues addressed by this PR

 Closes #296 

 

 ### Test files

https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/GSA_Toolkit/%23297-Pull%20and%20Push%20Constraints?csf=1&web=1&e=fDky6R

 ### Changelog

 ### Additional comments

### CODE REVIEW

File: GSA_oM/Elements/RigidConstraint.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+using System.Collections.Generic;
+using BH.oM.Quantities.Attributes;
+using BH.oM.Structure.Elements;
+using BH.oM.Adapters.GSA.SpacerProperties;
+using BH.oM.Analytical.Elements;
+
+namespace BH.oM.Adapters.GSA.Elements
+{
+    [Description("A rigid constraint defining linkage between nodes.")]
+    public class RigidConstraint : BHoMObject
```
Comment (IsakNaslundBh) [inline GSA_oM/Elements/RigidConstraint.cs:35]: Not to keen of the addition of this class is it fully overlaps, both in structure as well as intent, with the [RigidLink](https://github.com/BHoM/BHoM/blob/develop/Structure_oM/Elements/RigidLink.cs).

Another solution to this is to have RigidConstraints as well as Link elements both being read when asking for RigidLinks.

Can add a Fragment or similar in GSA_oM highlighting the type they actually refer to, and that can also help when pushing, dispatching out either to create a link element or RigidConstraint in GSA.

Probably warrants a call through how to handle this.

**Review by IsakNaslundBh — CHANGES_REQUESTED** [review]
See comment below. Think good to have a call to chat through

File: GSA_oM/Elements/RigidConstraint.cs (hunk 1 above)
Comment (peterjamesnugent) [reply GSA_oM/Elements/RigidConstraint.cs:35]: Based on the code, there are no new elements introduced that match the `RigidLink` and the fragment approach appears to have been adopted.

Comment (peterjamesnugent) [reply GSA_oM/Elements/RigidConstraint.cs:35]: This appears to have been removed when looking at the code but still comes up in the files changes:
image

**Comment by peterjamesnugent** [comment]
I have reviewed using the test script for 8.7 and 10.1, the Read works fine when the `RigidLink` (both Links and Cosntraints) have 1 node : 1 node relationships - the push also works for both.

However, when pushing a `RigidLink` (that is a GSA Link) with 1:n relationship, it needs to create several Rigid Links with the same name - this is problematic when pulling back as the `Node`s are not pulled correctly.

**Review by peterjamesnugent — APPROVED** [review]
I have tested using the test script for both GSA 10.7 and 8.1 - I have had to made two minor changes for it to function correctly (and to minimise altering the intent of the original PR):
- Change the delimiter from `','` to `' '` when pulling RigidConstraints from GSA that have a 1:many relationship
- Change how secondary nodes are handled for pushing RigidLinks to GSA 

The one issue I do have is pulling `Nodes` from a model, using them to create a `RigidLink`, creating a New File and then pushing the `RigidLink`. The `Node`s are not assigned properly - this may be because of the `GSAId` not being able to force a number - outside the scope of this PR and not sure whether that is a desire from a user point of view.

**Comment by Artem-Holstov** [comment]
**Review summary:** No new issues found in tests. Rigid links and rigid constraints push and pull functions work as expected.

**Comments:** GSA8.7 does not start automatically when the adapter is activated. The start model needs to be open in GSA separately before activating the adapter. In the initial tests, pushed rigid constraints did not appear in GSA. After rerunning the same process, the push function worked as expected. GSA8.7 was generally very unstable and prone to crashing in the tests. Hence, the initial glitch with pushing the RCs may be attributed to GSA8.7 and not the updated BHoM code. None of the above issues were observed with GSA 10.1.

**Features tested:**
Pull existing -> push/update -> pull updated functions tested for rigid links and rigid constraints with GSA 10.1 and 8.7 using test script. 
Pushed and pulled data was cross-checked between GH and GSA and logged in Excel.
Note the location of rigid links in the starting models was different between the files tested in GSA10.1 and GSA8.7. The script functions as expected with both starting models.

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance
@BHoMBot check required

**Comment by FraserGreenroyd** [comment]
@BHoMBot check ready-to-merge

**Review by FraserGreenroyd — APPROVED** [review]
Approving based on @peterjamesnugent and @Artem-Holstov reviews.

**Pull request merged**